master/pkg/searcher/asha.go

```diff
@@ -0,0 +1,244 @@
+	// number of best trials that definitely should have been promoted so far (numPromote) can only
+	// stay the same or increase by 1.
+	oldNumPromote := int(float64(len(r.metrics)) / divisor)
+	numPromote := int(float64(len(r.metrics)+1) / divisor)
+
+	// Insert the new trial result in the appropriate place in the sorted list.
+	insertIndex := sort.Search(
+		len(r.metrics),
+		func(i int) bool { return r.metrics[i].metric > metric },
+	)
+	promoteNow := insertIndex < numPromote
+
+	r.metrics = append(r.metrics, trialMetric{})
+	copy(r.metrics[insertIndex+1:], r.metrics[insertIndex:])
+	r.metrics[insertIndex] = trialMetric{
+		requestID: requestID,
+		metric:    metric,
+		promoted:  promoteNow,
+		closed:    false,
+	}
+
+	// If the new trial is good enough, it should be promoted immediately (whether or not numPromote
+	// changes). Otherwise, if numPromote changes, there is some other trial that should be promoted,
+	// unless it has been promoted already.
+	if promoteNow {
+		return []RequestID{requestID}
+	}
+	if numPromote != oldNumPromote {
+		t := &r.metrics[oldNumPromote]
+		if !t.promoted {
+			t.promoted = true
+			return []RequestID{t.requestID}
+		}
+	}
+
+	return nil
+}
+
+func (s *asyncHalvingSearch) initialOperations(ctx context) ([]Operation, error) {
+	// The number of initialOperations will control the degree of parallelism
+	// of the search experiment since we guarantee that each validationComplete
+	// call will return a new train workload until we reach MaxTrials.
+
+	// We will use searcher config field if available.
+	// Otherwise we will default to a number of trials that will
+	// guarantee at least one trial at the top rung.
+	var ops []Operation
+	var maxConcurrentTrials int
+
+	if s.MaxConcurrentTrials != nil {
+		maxConcurrentTrials = min(*s.MaxConcurrentTrials, s.MaxTrials)
+	} else {
+		maxConcurrentTrials = max(
+			min(int(math.Pow(s.Divisor, float64(s.NumRungs-1))), s.MaxTrials),
+			1)
+	}
+
+	for trial := 0; trial < maxConcurrentTrials; trial++ {
+		create := NewCreate(
+			ctx.rand, sampleAll(ctx.hparams, ctx.rand), model.TrialWorkloadSequencerType)
+		ops = append(ops, create)
+		ops = append(ops, trainAndValidate(create.RequestID, 0, s.rungs[0].stepsNeeded)...)
+		s.rungs[0].outstandingTrials++
+		s.trialRungs[create.RequestID] = 0
+	}
+	return ops, nil
+}
+
+func (s *asyncHalvingSearch) trainCompleted(
+	ctx context, requestID RequestID, message Workload,
+) ([]Operation, error) {
+	s.totalStepsCompleted++
+	return nil, nil
+}
+
+func (s *asyncHalvingSearch) validationCompleted(
+	ctx context, requestID RequestID, message Workload, metrics ValidationMetrics,
+) ([]Operation, error) {
+	// Extract the relevant metric as a float.
+	metric, err := metrics.Metric(s.AsyncHalvingConfig.Metric)
+	if err != nil {
+		return nil, err
+	}
+	if !s.AsyncHalvingConfig.SmallerIsBetter {
+		metric *= -1
+	}
+
+	return s.promoteAsync(ctx, requestID, message, metric), nil
+}
+
+func (s *asyncHalvingSearch) promoteAsync(
```

Comment:
non-blocking: for consistency we should change this to `promote` and `promotionsAsync` -> `promotions`. It should be clear from that this is called `asyncHalvingSearch` that these are async

Reply:
i originally had them named that but then go complained that promotions was redeclared since it operates on rungs which is shared by sha and asha.  do you think it's better to have (s *asyncHalvingSearch) promote or (s *asyncHalvingSearch) promoteAsync accompany  (r * rung) promoteAsync?

Reply:
in sha we use `promoteTrials` right?

Reply:
yes, it's promoteTrials in sha. what do you think about promotionsSync and promoteSync  and promotionsAsync and promoteAsync?

Reply:
I can get behind that 👍